satellite/audit/service2.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// Service2 contains information for populating audit queue and processing audits.
+type Service2 struct {
+	log    *zap.Logger
+	closed chan struct{}
+
+	workers []*worker
+	queue   *queue
+}
+
+// NewService2 instantiates Service2 and workers.
+func NewService2(log *zap.Logger, config Config) (*Service2, error) {
+	queue := &queue{}
+	var workers []*worker
+	for i := 0; i < config.WorkerCount; i++ {
+		workers = append(workers, newWorker(queue, config.QueueInterval))
+	}
+	return &Service2{
+		log:    log,
+		closed: make(chan struct{}),
+
+		workers: workers,
+		queue:   queue,
+	}, nil
+}
+
+// Run runs audit service 2.0.
+func (service *Service2) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("audit 2.0 is starting up")
+
+	var group errgroup.Group
+	for _, worker := range service.workers {
+		group.Go(func() error {
+			return worker.run(ctx)
+		})
+	}
+
+	return group.Wait()
+
+}
+
+// Close halts the reservoir chore and audit workers.
+func (service *Service2) Close() error {
+	close(service.closed)
```

Comment:
I do not understand why there is this `closed` here?

Reply:
Oh, I don't think it got captured in the latest refactor, but we plan to check `service.closed` inside the loop inside `func (*worker) run(ctx context.Context)`. Alternatively, we could create a new ctx with cancel that we pass into that function, and cancel it here. Do you have a preference for how we handle this?

Reply:
https://github.com/storj/storj/blob/master/satellite/repair/repairer/repairer.go#L70

Reply:
We can do a similar approach as in repairer.

Reply:
I tried to make it similar to repairer, but not sure if it is what you had in mind

Reply:
https://github.com/storj/storj/pull/2924/files